### DISCUSSION TITLE
Having issues with ignores in v9

### DISCUSSION THREAD

**Discussion by sebnitu** [eslint/eslint#18304] [Q&A] [answered]
I'm in the process of migrating over to v9 and am having a hard time getting `ignores` to work as expected. My old implementation looked like this and it ignored these directories and all of their content:

```js
ignorePatterns: ['dev', 'dist', 'tests', 'tsconfig.json']
```

In the new flat config file, I converted this to look like:

```js
ignores: [
  '**/dev/*',
  '**/dist/*',
  '**/tests/*',
  'tsconfig.json',
],
```

but when I run eslint, I get a bunch of errors in files that should be getting ignored. Here are some of my testing to try and figure out how it works:

---

```js
ignores: []
```

Result: 7465 problems (7465 errors, 0 warnings)

---

```js
ignores: [
  '**/dev/*'
]
```

Result: 5159 problems (5159 errors, 0 warnings)
Ok, on the right track, lets keep going...

---

```js
ignores: [
  '**/dev/*',
  '**/dist/*'
]
```

Result: 2157 problems (2157 errors, 0 warnings)
Event better, keep going...

---

```js
ignores: [
  '**/dev/*',
  '**/dist/*',
  '**/tests/*'
]
```

Result: 2481 problems (2481 errors, 0 warnings)
Wait, what? Why did the number of errors go up after adding another ignore directory?!?

At this point, I can't get the errors to go down anymore and when I look at the first offending file, it's in this directory:

```
/Users/sebastian/dev/vrembem/packages/vrembem/dist/index.umd.cjs
```

Why is this file not being ignored with the `'**/dist/*'` rule?

Incase it matters, my `eslint.config.js` file is in the root of my project: `/Users/sebastian/dev/vrembem/eslint.config.js` and according to the docs that's where the globs should be starting at, correct?

**Comment by sebnitu** [comment]
For whatever reason removing `js.configs.recommended` from my config fixed my issue:

```js
import js from '@eslint/js';

export default [
  js.configs.recommended,
  {
    // ...
  }
]
```

I even copy/pasted all the rules from `https://www.npmjs.com/package/@eslint/js?activeTab=code` into my own config so all the same rules should be applied but it's no longer triggering on files that should be getting ignored. 🤷

**Comment by kewisch** [comment, marked as answer]
Have you tried putting `ignores` in a config object by itself? https://github.com/eslint/eslint/discussions/17429

```javascript
export default [
  {
    ignores: [
      '**/dev/*',
      '**/dist/*',
      '**/tests/*',
      'tsconfig.json',
    ]
  },
  js.configs.recommended,
  {
    // ...
  }
]
```

**Reply by sebnitu** [reply]
Yup, that fixed my issue! I'm still having a hard time wrapping my head around why this worked tho? I've read the docs a few times especially related to [`Configuration Objects`](https://eslint.org/docs/latest/use/configure/configuration-files#configuration-objects) but I still don't think I'm fully wrapping my head around how that works. Does every "config object" cause the linter to run except for these "special" objects that just act as globals? Why doesn't `js.configs.recommended` also act as a global as well instead of (what I think it's doing) running it's own linting along side my custom object?

Either way, thank you for the suggestion!

**Reply by kewisch** [reply]
I haven't fully understood it either yet, but here is what I know:
* The `ignores` by itself is a specical construct to apply global ignores. If you add any other keys, it turns into a normal block and behaves differently.
* Each subsequent block is applied in a cascading manner.
  * So js.configs.recommended provides a set of rules. They will apply to all future config blocks.
  * If you have a next block that has e.g. `files` in it, it will take whatever came before that, and apply the new rules on top, but only for the files specified. This will propagate to other blocks in the future that affect the same files.
  * The same goes for `ignores` with other keys. It will apply the new rules to all files, except those in the ignores section.
  * If you have another block after this that doesn't have ignores, it will take the rules from prior blocks where the files were within the selection.


I can imagine that the number went up because the ignores block did its job ignoring the rules, but a subsequent block was then checking the supposedly ignored files with a different combination of enabled rules.

**Reply by davestewart** [reply]
@kewisch you've probably saved me about $1000 in replacement laptop this evening, as mine was about to be thrown at the wall. Thanks!

**Reply by kewisch** [reply]
I take donations ;-)

**Reply by davestewart** [reply]
I'll buy you a pint next time I see you!

**Reply by alvaropsouza** [reply]
Yes, you actually saved me from purging eslint from my project. Thank you.

**Reply by felipemullen** [reply]
I just spent 2 hours trying to understand the very same problem with the ignores. 
This should definitely be properly documented, as going through `eslint` and `typescript-eslint` docs do not have anything more than simple starter configs to look at.
The moment you dive into a real project (migrating a react app in my case), you're left with all these code samples on the internet that are using the old config style

**Reply by gitmp01** [reply]
yeah totally agree here, this should be noted in capital letters in the docs

**Reply by wporoslo** [reply]
jesus christ, thank you!!!!!!!!!!

**Reply by michael-harley** [reply]
Thanks! This solution still work for 9.7.0.

**Reply by tserencio-fs** [reply]
Been cracking on this issue for 2 days already. Thanks man! You saved me

**Reply by AndrewStobie** [reply]
this worked for me too!

**Reply by Danielvandervelden** [reply]
Bro why the hell is this not in font size 200rem in the docs...

**Reply by Joecey** [reply]
Amazing thank you! For anyone using `typescript-eslint` recommended linter, you can expand it with the ignored folders as such

Why this isn't mentioned as the first thing on the docs is beyond me 

```
import eslint from '@eslint/js'
import tseslint from 'typescript-eslint'

export default [
    {
        ignores: ['**/build/*', '**/node_modules/*'],
    },
    eslint.configs.recommended,
    ...tseslint.configs.recommended,
]

```

**Reply by hivokas** [reply]
Thank you! I think it should be emphasized in docs a bit more 🤣 

<img width="658" alt="Screenshot 2024-09-02 at 5 00 52 PM" src="https://github.com/user-attachments/assets/6cec5e9b-9f4a-46ab-9302-5e3b03922e57">

**Reply by murtuzaalisurti** [reply]
well, this should also have an actual code example in the docs, was scratching my head for a long time!

**Reply by TokugawaTakeshi** [reply]
In means, below example in [docs](https://eslint.org/docs/latest/use/configure/migration-guide#ignoring-files) is inaccurate and requires the additional explanation?

```
export default [
    // ...other config
    {
        ignores: ["**/temp.js", "config/*"]
    }
];
```

**Reply by mystiker** [reply]
Jesus thank you very much guys. You saved me from insanity (till tomorrow).

**Reply by seanpmaxwell** [reply]
Finally!

**Reply by ceoworks** [reply]
it's also possible to have a structure like this:

```typescript
export default tseslint.config(
  {
    // Extend ESLint's recommended rules
    extends: [
      eslint.configs.recommended,
      ...tseslint.configs.recommended,
    ],
    
    // Project files to lint
    files: ["**/*.ts"],
    
    // Ignore patterns
    ignores: [
      "node_modules/**",
      "dist/**",
      "build/**",
      ".next/**",
      "coverage/**",
    ],
```

**Reply by YornQiu** [reply]
Amazing! Thanks God!

**Reply by steida** [reply]
This is pain. It's bad it's not documented, and Eslint keeps throwing weird errors instead. Eslint should Eslint itself.

**Reply by nathan-chappell** [reply]
I'm already sorry I asked, but does anyone know why it's like this?  Eslint is a massively popular tool, and it seems like a decent number of people have issues with this.  Like, putting *ignore* in the config has no effect, but putting it in it's own object has the desired effect - that's not intuitive at all and is in fact quite surprising.

**Reply by leviFrosty** [reply]
I spent too ******** long trying to nail down this issue, all to be some flattening problem on ESlint's side. This is awful.

**Reply by pfroud** [reply]
Here is a closed pull request to emphasize that `ignore` must be put in its own object:

https://github.com/eslint/eslint/pull/18963

**Reply by mitsukuri** [reply]
Seriously, this is the most idiotic thing I encountered in coding in 2024

**Reply by melroy89** [reply]
I hate the flat config so much

**Reply by tiagocardosoc** [reply]
I was dealing with this error all the morning, thanks for the solution. But this flat config is TERRIBLE.

**Reply by ThibaultWalterspieler** [reply]
You're the best, I've spent hours on this.  ❤️

**Reply by thefaded** [reply]
The most awful configuration I have ever encountered. The 9th version of ESLint is simply total trash. There is a feeling that it was released merely to give a semblance of work for OSS donations.

**Reply by alan890104** [reply]
presonally really don't like the design, this is unfriendly. Thanks for the solution

**Reply by applepie1223** [reply]
thank you!

**Reply by TobiWo** [reply]
OMG, this solved the issue for me as well. Thank you so so much. I agree with all the replies above: The flat config design is not really dev friendly. Anyway, I hope I get used to it. Seems like the only way.

**Reply by krutoo** [reply]
Thank you, but... This is the most strange and inappropriate thing in the ESLint 9, this is completely unintuitive, this is literally the worst configuration design I have ever seen

**Reply by bert-w** [reply]
I have decided to stay on an older version of eslint. This whacky config is unexplainable to me even though it is a little tidier.

**Comment by ant1m4tt3r** [comment]
had the same issue but I guess my case is a bit different. what if I want to ignore files just based on a single rule? like for instance:

```js
 {
    files: ['**/*.ts', '**/*.tsx'],

    ignores: [
      'src/vite-env.d.ts',
      'src/lib/some-file.ts',
    ],

    rules: {
      '@typescript-eslint/no-explicit-any': 'error',
    },
  },
```

in this case, I wan't to ignore the `@typescript-eslint/no-explicit-any` rule only for these files. if i let the config like I described above eslint won't ignore them.

My whole config for comparisson:

```js
import path from 'node:path'
import { fileURLToPath } from 'node:url'
import js from '@eslint/js'
import { FlatCompat } from '@eslint/eslintrc'

const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)
const compat = new FlatCompat({
  baseDirectory: __dirname,
  recommendedConfig: js.configs.recommended,
  allConfig: js.configs.all,
})

export default [
  {
    ignores: ['**/mockServiceWorker.js'],
  },
  {
    files: ['**/*.ts', '**/*.tsx'],

    ignores: [
      'src/vite-env.d.ts',
      'src/lib/some-file.ts',
      'tests/lib/other-file.spec.tsx',
      'src/lib/other-file.tsx',
    ],

    rules: {
      '@typescript-eslint/no-explicit-any': 'error',
    },
  },
  ...compat.extends('../../../.eslintrc.json'),
]
```

**Reply by mdjermanovic** [reply]
> ```js
> export default [
>   {
>     ignores: ['**/mockServiceWorker.js'],
>   },
>   {
>     files: ['**/*.ts', '**/*.tsx'],
> 
>     ignores: [
>       'src/vite-env.d.ts',
>       'src/lib/some-file.ts',
>       'tests/lib/other-file.spec.tsx',
>       'src/lib/other-file.tsx',
>     ],
> 
>     rules: {
>       '@typescript-eslint/no-explicit-any': 'error',
>     },
>   },
>   ...compat.extends('../../../.eslintrc.json'),
> ]
> ```

Do you have `@typescript-eslint/no-explicit-any` enabled in your `.eslintrc.json` file?

**Comment by Jakobud** [comment]
The `ignores` configuration implementation is one of the most unintuitive and inconsistent things I have seen in a long time.

The documentation shows this example:

```
export default [
    {
        files: ["src/**/*.js"],
        ignores: ["**/*.config.js"],
        rules: {
            semi: "error"
        }
    }
];
```

and also says 

> Non-global ignores patterns can only match file names. A pattern like "dir-to-exclude/" will not ignore anything. To ignore everything in a particular directory, a pattern like "dir-to-exclude/**" should be used instead.

So I try this:

```
import js from "@eslint/js";

export default [
  js.configs.recommended,
  {
    files: ["src/**/*.ts"],
    ignores: ["dist/**/*.js"],
    languageOptions: {
      ecmaVersion: 2022,
      sourceType: "module",
    },
  },
];
```

and of course this doesn't work. But then if I remove `js.config.recommended`:

```
export default [
  {
    files: ["src/**/*.ts"],
    ignores: ["dist/**"],
    languageOptions: {
      ecmaVersion: 2022,
      sourceType: "module",
    },
  },
];
```

TADA now it works. WTF kind of magic bullshit is this? How is anyone supposed to figure this out? Either the implementation sucks or the documentation sucks. Why is it so complicated to implement a stupid glob pattern matcher?

And it seems like everyone's solution is to say "just move your ignores to it's own configuration object and it fixes it". What? Why does that fix it? What happens if I have multiple configuration objects and I need to ignore something only for a specific config?

Sorry but this is terrible.

**Reply by bloudermilk** [reply]
While I can't abide your tone toward a free tool maintained by gracious contributors, I can attest that the specific issue you're describing is confounding and frustrating. I'm hopeful the ignore interface will get refined based on the feedback in this thread.

In my case, the following config worked among the rest of my rules:

```ts
  // console rules for ES/TS across all workspaces
  {
    files: ['**/*.{ts,js,vue,mjs,cjs}'],
    ignores: ['app/tests/**', '**/*.stories.*', 'api/**'],
    rules: {
      'no-console': ['error', { allow: ['warn', 'error', 'debug'] }]
    }
  },
```

**Comment by kewisch** [comment]
The `ignores` by itself is a global ignore for all configs. You can ignore per-entry by including `ignores` like you have. What happens with your example though is that `js.configs.recommended` does not have `ignores`, so _that one_ will apply to all files, but your second entry will ignore dist files. So you'd need this (untested):

```
import js from "@eslint/js";

export default [
  {
  ...js.configs.recommended,
  ignores: ["dist/**/*.js"]
  },
  {
    files: ["src/**/*.ts"],
    ignores: ["dist/**/*.js"],
    languageOptions: {
      ecmaVersion: 2022,
      sourceType: "module",
    },
  },
];
```

The documentation could be more clear on this. I personally think the config itself is very capable, but it takes a little bit to wrap your head around.

**Reply by michael-harley** [reply]
So... By moving the ignore to the top array, we basically will ignore all of the other config (including the one from the recommended config)
cmiiw.

**Comment by yunnysunny** [comment]
I found the solution to the problem: https://github.com/eslint/eslint/issues/17400#issuecomment-1647463901

**Comment by nureon22** [comment]
If I specified more than one object in exported array, I need to put `ignores` property with exact same value in every object inside this array. Just using one object work fine.

```
import globals from "globals";
import pluginJs from "@eslint/js";

export default [
    {
        languageOptions: { globals: globals.browser, ecmaVersion: 5, sourceType: "script" },
        ignores: ["static/**/*", "public/**/*", "postcss.config.js", "eslint.config.mjs"],
        rules: pluginJs.configs.recommended.rules,
    },
];
```

**Comment by pcone** [comment]
The way this works currently is really unintuitive, even adding 
`settings: {
  reportUnusedDisableDirectives: true,
},`
To the same object as my ignores directive seems to stop it from applying - surely this is not intended behaviour?

**Comment by stefanseeger** [comment]
I know this discussion is cold now.

But I would like to add that I recognized that when I run `npx @eslint/config-inspector` and go to the **Files** tab I see less files (ignored seems to be processed differently) than are actually linted when I run `npx eslint .`

**Comment by EmilTholin** [comment]
**Edit:** Comment by @egor-xyz on this reply is better.

---

Since [`tseslint.config` is just a utility for creating an array of configs](https://typescript-eslint.io/packages/typescript-eslint/#flat-config-extends) we can spread the result of that into an array of our own where we tell eslint to ignore certain files for all configs.

**Example**

```js
import js from '@eslint/js';
import globals from 'globals';
import reactHooks from 'eslint-plugin-react-hooks';
import tseslint from 'typescript-eslint';

export default [
  { ignores: ['storybook-static/**'] },
  ...tseslint.config({
    extends: [js.configs.recommended, ...tseslint.configs.recommended],
    files: ['**/*.{ts,tsx}'],
    languageOptions: {
      ecmaVersion: 2020,
      globals: globals.browser,
    },
    plugins: {
      'react-hooks': reactHooks,
    },
    rules: {
      ...reactHooks.configs.recommended.rules,
      '@typescript-eslint/no-explicit-any': 'off',
    },
    languageOptions: {
      parserOptions: {
        warnOnUnsupportedTypeScriptVersion: false,
      },
    },
  }),
];
```

**Reply by egor-xyz** [reply]
I just reccomend to wrap all:
```js
import js from '@eslint/js';
import globals from 'globals';
import reactHooks from 'eslint-plugin-react-hooks';
import tseslint from 'typescript-eslint';

export default tseslint.config(
  { ignores: ['storybook-static/**'] },
  {
    extends: [js.configs.recommended, ...tseslint.configs.recommended],
    files: ['**/*.{ts,tsx}'],
    languageOptions: {
      ecmaVersion: 2020,
      globals: globals.browser,
    },
    plugins: {
      'react-hooks': reactHooks,
    },
    rules: {
      ...reactHooks.configs.recommended.rules,
      '@typescript-eslint/no-explicit-any': 'off',
    }
  },
);
```

**Comment by steida** [comment]
Here is a piece of "flat-config" that works. It took me hours to write it. Still, I'm not sure why it works. Hope it helps.

https://github.com/eslint/eslint/issues/19093#issuecomment-2453578291

**Comment by Houuse** [comment]
after wasting many hours I got this to _work_, it would be better to remove this useless tool altogether, which I will do in the next MR
For now for those who comes here looking for help, 
it turned out I needed 2 different config blocks for it to _work_
like this:
`    {
        ignores: ["jest.config.test.js", "babel.config.js", "something/**/*.js", "sthing-sthing/**/*.js"],
    },
   {
      // other config
   }
`

**Comment by tresorama** [comment]
I still don't understand how this ignore props is supposed to be used when the eslint config is an array of multiple config, but i found that:

if in the default exported array of the config you define the `ignores` config in the first item of the array, **AND YOU DON't ADD ANY OTHER PROPERTY TO THE SAME OBJECT** the `ignores` works as supposed to

```js
import globals from "globals";
import pluginJs from "@eslint/js";
import tseslint from "typescript-eslint";


/** @type {import('eslint').Linter.Config[]} */
export default [
  // DO NOT PUT OTHER PROPS IN THIS OBJECT
  { ignores: ["**/node_modules/**", "dist/"] }, 

  // THE IGNORES IS APPLIED ALSO TO THESE FOLLOWING CONFIGS
  { files: ["**/*.{js,mjs,cjs,ts}"] },
  { languageOptions: { globals: globals.browser } },
  pluginJs.configs.recommended,
  ...tseslint.configs.recommended,
];
```

**Reply by jpbnetley** [reply]
Thank you for this, this solved my issue where the `ignores` field was... ignored. 
How ironic.
I must say, using the eslint flat config is quite... tricky for me :)

**Reply by cvazquez** [reply]
That solved it for me too. 

I was trying to ignore files with a \*copy\*.js pattern, which get created when you copy/paste a file within a directory in Visual Studio Code. Adding it to a separate object at the top worked for me:
`{
      ignores: ["**/*/*copy*.js"],
},`

Thank you!

**Comment by SerkanSipahi** [comment]
@tresorama Thanks. Ignoring all other properties when using `ignores` solved the issue. One observation for all others: Dont start path patterns with dot slash  `./whatever/path/**/*` (does not work) instead, use just `whatever/path/**/*`.

@nzakas would you be so kind as to explain this (see the thread above) to us and possibly also show us the section in the documentation where this is described. I think some people here are a bit frustrated but I think this can easily be cleared up by you :) Many thanks in advance.

**Reply by tresorama** [reply]
When people search for "how to ignore files in eslint" ideally they land on [this docs pages](https://eslint.org/docs/latest/use/configure/configuration-files#configuration-objects).

The `ignores` description partially-include the strange behavior

> ignores - An array of glob patterns indicating the files that the configuration object should not apply to. If not specified, the configuration object applies to all files matched by files. If ignores is used without any other keys in the configuration object, then the patterns act as [global ignores](https://eslint.org/docs/latest/use/configure/configuration-files#globally-ignoring-files-with-ignores).

But currently indicate the reader to click on `global ignores` link.
The [landing page](https://eslint.org/docs/latest/use/configure/configuration-files#globally-ignoring-files-with-ignores) has a not-enough description. If i read it now it don't understand how to use, because of:
- text not mentioning "if you want a unique place to define which files/dirs eslint must ignore you must add a single config object as **first item** of all configs of `eslint.config.*` file, and **do not add any other props to the same config object**. Like in this snippets..."
- snippet is not understandable, because doesn't clearify that the order of config object matters. Now the snippets has a single object, but to understand the behavior the snippets should have multiple configs.

I think that one of the first things a user of eslint want to do is to filter out some directories/files, so a copy paste snippets of code is a MUST HAVE.

**Reply by nzakas** [reply]
Please feel free to open a PR with the changes to the docs you think would be helpful. There's an "Edit This Page" button at the bottom of every page that can help get you started.

**Reply by tresorama** [reply]
> Please feel free to open a PR with the changes to the docs you think would be helpful. There's an "Edit This Page" button at the bottom of every page that can help get you started.

Opened PR https://github.com/eslint/eslint/pull/19297
